packages/mobile/src/account/profileInfo.ts

```diff
@@ -0,0 +1,326 @@
+import { call, put, select } from 'redux-saga/effects'
+import { profileUploaded } from 'src/account/actions'
+import { isProfileUploadedSelector, nameSelector, pictureSelector } from 'src/account/selectors'
+import config from 'src/geth/networkConfig'
+import { extensionToMimeType, getDataURL, saveRecipientPicture } from 'src/utils/image'
+import Logger from 'src/utils/Logger'
+import { getContractKit, getWallet } from 'src/web3/contracts'
+import { getConnectedUnlockedAccount } from 'src/web3/saga'
+import { currentAccountSelector, dataEncryptionKeySelector } from 'src/web3/selectors'
+
+const TAG = 'account/profileInfo'
+
+async function authorizeURLs(data: any, signature: string): Promise<SignedPostPolicyV4Output[]> {
+  const response = await fetch(config.CIP8AuthorizerUrl, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      Signature: signature,
+    },
+    body: JSON.stringify(data),
+  })
+
+  if (response.status >= 400) {
+    throw new Error(await response.text())
+  }
+
+  return response.json()
+}
+
+// Workaround fetch response.arrayBuffer() not working in RN environment
+// See https://github.com/facebook/react-native/blob/f96478778cc00da8c11da17f9591dbdf928e7437/Libraries/Blob/FileReader.js#L85
+async function responseBuffer(response: Response) {
+  const blob = await response.blob()
+  return blobToBuffer(blob)
+}
+
+// Hacky way to get Buffer from Blob
+// Note: this is gonna transfer the whole data over the RN bridge (as base64 encoded string)
+// and should be avoided for large files!
+function blobToBuffer(blob: Blob) {
+  const reader = new FileReader()
+  reader.readAsDataURL(blob)
+  return new Promise<Buffer>((resolve, reject) => {
+    reader.onerror = () => {
+      reject(reader.error)
+    }
+    reader.onloadend = () => {
+      const result = reader.result
+      if (typeof result !== 'string') {
+        reject(new Error('Unexpected result type'))
+      } else {
+        // Result looks like "data:application/octet-stream;base64,BLMHTkM..."
+        // Extract the base64 part
+        const base64 = result.substr(result.lastIndexOf(',') + 1)
+        resolve(Buffer.from(base64, 'base64'))
+      }
+    }
+  })
+}
+
+class UploadServiceDataWrapper implements OffchainDataWrapper {
+  signer: Address
+  self: Address
+
+  constructor(readonly kit: ContractKit, address: Address) {
+    this.signer = this.self = address
+  }
+
+  async writeDataTo(
+    data: Buffer,
+    signature: Buffer,
+    dataPath: string
+  ): Promise<OffchainErrors | void> {
+    const dataPayloads = [data, signature]
+    const signedUrlsPayload = [
+      {
+        path: dataPath,
+      },
+      {
+        path: `${dataPath}.signature`,
+      },
+    ]
+
+    const hexPayload = ensureLeading0x(
+      Buffer.from(JSON.stringify(signedUrlsPayload)).toString('hex')
+    )
+    const authorization = await this.kit.getWallet().signPersonalMessage(this.signer, hexPayload)
```

Comment:
It seems that this would require the account key to be unlocked to complete this action. If the account key is not unlocked, I believe this will throw an error. Is this handled?

Also, it would be better if this message was signed using an EIP-712 structure rather than hex encoded JSON. It's not critical, but it would definitly be easier to change this now if possible rather than later.

Reply:
yes, `yield call(getConnectedUnlockedAccount)` ensures that the account is unlocked